src/nvim/event/process.c

```diff
@@ -227,30 +228,26 @@ void process_stop(Process *proc) FUNC_ATTR_NONNULL_ALL
       abort();
   }
 
-  Loop *loop = proc->loop;
-  if (!loop->children_stop_requests++) {
-    // When there's at least one stop request pending, start a timer that
-    // will periodically check if a signal should be send to the job.
-    ILOG("starting job kill timer");
-    uv_timer_start(&loop->children_kill_timer, children_kill_cb,
-                   KILL_TIMEOUT_MS, KILL_TIMEOUT_MS);
-  }
+  // Start a timer to verify that the job process terminated.
+  ILOG("starting job kill timer");
+  uv_timer_start(&proc->loop->children_kill_timer, children_kill_cb,
+                 KILL_TIMEOUT_MS, 0);
 }
 
-/// Iterates the process list sending SIGTERM to stopped processes and SIGKILL
-/// to those that didn't die from SIGTERM after a while(exit_timeout is 0).
+/// Sends SIGKILL (or SIGTERM for PTY jobs) to processes that didn't terminate
+/// after process_stop() requested them.
 static void children_kill_cb(uv_timer_t *handle)
 {
   Loop *loop = handle->loop->data;
   uint64_t now = os_hrtime();
 
   kl_iter(WatcherPtr, loop->children, current) {
     Process *proc = (*current)->data;
-    if (!proc->stopped_time) {
+    bool exited = (proc->status >= 0);
+    if (exited || !proc->stopped_time) {
       continue;
     }
     uint64_t elapsed = (now - proc->stopped_time) / 1000000 + 1;
-
     if (elapsed >= KILL_TIMEOUT_MS) {
       int sig = proc->type == kProcessTypePty && elapsed < KILL_TIMEOUT_MS * 2
```

Comment:
Should this code be changed so that always a `SIGKILL` is send?
* Calling `process_stop()` for a pty that ignores `SIGTERM` would probably never on its own reach an elapsed time of `2*KILL_TIMEOUT_MS`. The `pty` would have to wait until another process is stoped so that the kill timer is restarted.
* If the timer fires too early no signal could be send and the process would again had to wait for another stopped process.

Reply:
@oni-link Good catch again. @blueyed was there a specific reason we need to try SIGTERM before SIGKILL, if a PTY process didn't terminate after SIGHUP?

Reply:
Is not uncommon for there to be special handling of `SIGHUP`, e.g. to reload config or show progress.  `SIGTERM`, on the other hand, is unlikely to have an overloaded meaning.  Therefore issuing `SIGTERM` still gives an opportunity for graceful shutdown before the sledgehammer of `SIGKILL`.